### PR TITLE
Emotions on posts

### LINKED ISSUE #384: 🚀 [Feature] Emotion Reactions on Posts
> [ulfgebhardt](https://github.com/ulfgebhardt) **Authored by [ulfgebhardt](https://github.com/ulfgebhardt)**
_<time datetime="2019-04-03T15:57:02Z" title="Wednesday, April 3rd 2019, 5:57:02 pm +02:00">Apr 3, 2019</time>_
_Closed <time datetime="2019-08-08T22:37:39Z" title="Friday, August 9th 2019, 12:37:39 am +02:00">Aug 9, 2019</time>_
---

### Emotion Reactions on Posts
image

- [ ] Emotions
- [ ] Image active + inactive @appinteractive can you provide the images?
- [ ] ~Percentage~ Absolute values by Userreaction distribution
- [ ] User Interaction (click/click again -> change)

> HC users wish better filter possibilities in order to suppress "annoying" user and usercontent from their personal Newsfeed. Extending the range of emojis from 5-10 is a good and easy way to create more comfort.

Emotions:
- Funny
- Happy
- Surprised
- Cry
- Angry
- Love (heart)
- Disconnected (broken heart)
- fear (bibbernd)
- questioning (questioning emoji)
- resignated (hand touching face)
- help! (siren on head feat. Minion :-D)  

Note: Once selected deselection is not possible

### Userstory
```
As a User
  I`d like posts to be tagged with emotions
  in order to be able to filter them by such
```

### Design & Layout - Alpha

No User Selection:
image

With User Selection:
image

### Design & Layout - Nitro
missing

### PR DESCRIPTION
**Pull request by Tirokk**
> [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-07-29T15:09:00Z" title="Monday, July 29th 2019, 5:09:00 pm +02:00">Jul 29, 2019</time>_
_Merged <time datetime="2019-08-08T22:37:39Z" title="Friday, August 9th 2019, 12:37:39 am +02:00">Aug 9, 2019</time>_
---

### Issues

- fixes #384

### CODE REVIEW

**Comment by Tirokk** [comment]
> [appinteractive](https://github.com/appinteractive) **Authored by [appinteractive](https://github.com/appinteractive)**
_<time datetime="2019-08-01T21:58:29Z" title="Thursday, August 1st 2019, 11:58:29 pm +02:00">Aug 1, 2019</time>_

---


@mattwr18 I love the term „emoted“, you made my day. 😅 is it Even s Word?

**Comment by Tirokk** [comment]
> [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-02T05:38:52Z" title="Friday, August 2nd 2019, 7:38:52 am +02:00">Aug 2, 2019</time>_

---


That's great @appinteractive, I'm happy to have made your day :smile_cat: 
especially, when I didn't purposely set out to :stuck_out_tongue: 
yes, it's the past tense of emote :rofl:

**Comment by Tirokk** [comment]
> [appinteractive](https://github.com/appinteractive) **Authored by [appinteractive](https://github.com/appinteractive)**
_<time datetime="2019-08-02T15:17:07Z" title="Friday, August 2nd 2019, 5:17:07 pm +02:00">Aug 2, 2019</time>_

---


Learned something again 😌

File: backend/src/schema/resolvers/posts.spec.js [hunk 1]
```diff
@@ -383,3 +400,315 @@ describe('DeletePost', () => {
...
+          user,
+          removePostEmotionsMutation,
+          removePostEmotionsVariables,
+        )
+        expect(removePostEmotions.errors[0]).toHaveProperty('message', 'Not Authorised!')
+      })
+    })
+
+    describe('authenticated', () => {
+      describe('but not the emoter', () => {
```
Comment (Tirokk) [inline backend/src/schema/resolvers/posts.spec.js:626]: > [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-08-05T14:22:04Z" title="Monday, August 5th 2019, 4:22:04 pm +02:00">Aug 5, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -383,3 +385,220 @@ describe('DeletePost', () => {
     })
   })
 })
+
+describe('emotions', () => {
+  let addPostEmotionsVariables
+  let postEmotionsQueryVariables
+  const postEmotionsQuery = `
+    query($id: ID!) {
+      Post(id: $id) {
+        emotions {
+          emotion
+          User {
+            id
+          }
+        }
+      }
+    }
+  `
+  beforeEach(async () => {
+    const asAuthor = Factory()
+    authorParams = {
+      id: authorParams.id,
+      email: 'wanna-add-emotions@example.org',
+      password: '1234',
+    }
+    await asAuthor.create('User', authorParams)
+    await asAuthor.authenticateAs(authorParams)
+    await asAuthor.create('Post', {
+      id: 'p1376',
+      title: postTitle,
+      content: postContent,
+    })
+    postEmotionsQueryVariables = { id: 'p1376' }
+  })
+
+  describe('AddPostEmotions', () => {
+    beforeEach(() => {
+      addPostEmotionsVariables = {
+        from: { id: authorParams.id },
+        to: { id: 'p1376' },
+        data: { emotion: 'happy' },
+      }
+    })
+
+    const addPostEmotionsMutation = `
+      mutation($from: _UserInput!, $to: _PostInput!, $data: _EMOTEDInput!) {
+        AddPostEmotions(from: $from, to: $to, data: $data) {
+          from { id }
+          to { id }
+          emotion
+        }
+      }
+    `
+    const postEmotionsCountQuery = `
+      query($id: ID!) {
+        Post(id: $id) {
+          emotionsCount
+        }
+      }
+    `
+
+    describe('unauthenticated', () => {
+      it('throws authorization error', async () => {
+        client = new GraphQLClient(host)
+        await expect(
+          client.request(addPostEmotionsMutation, {
+            from: { id: authorParams.id },
+            to: { id: 'p1376' },
+            data: { emotion: 'happy' },
+          }),
+        ).rejects.toThrow('Not Authorised')
+      })
+    })
+
+    describe('authenticated and not the author', () => {
+      let headers
+      beforeEach(async () => {
+        headers = await login(userParams)
+        client = new GraphQLClient(host, { headers })
+      })
+
+      it('adds an emotion to the post', async () => {
+        addPostEmotionsVariables.from.id = userParams.id
+        const expected = {
+          AddPostEmotions: {
+            from: addPostEmotionsVariables.from,
+            to: addPostEmotionsVariables.to,
+            emotion: 'happy',
+          },
+        }
+        await expect(
+          client.request(addPostEmotionsMutation, addPostEmotionsVariables),
+        ).resolves.toEqual(expected)
+      })
+
+      it('limits the addition of the same emotion to 1', async () => {
+        const expected = {
+          Post: [
+            {
+              emotionsCount: 1,
+            },
+          ],
+        }
+        await client.request(addPostEmotionsMutation, addPostEmotionsVariables)
+        await client.request(addPostEmotionsMutation, addPostEmotionsVariables)
+        await expect(
+          client.request(postEmotionsCountQuery, postEmotionsQueryVariables),
+        ).resolves.toEqual(expected)
+      })
+
+      it('allows a user to add more than one emotion', async () => {
+        const expected = [
+          { emotion: 'happy', User: { id: authorParams.id } },
+          { emotion: 'surprised', User: { id: authorParams.id } },
+        ]
+        await client.request(addPostEmotionsMutation, addPostEmotionsVariables)
+        addPostEmotionsVariables.data.emotion = 'surprised'
+        await client.request(addPostEmotionsMutation, addPostEmotionsVariables)
+        await expect(
+          client.request(postEmotionsQuery, postEmotionsQueryVariables),
+        ).resolves.toEqual({
+          Post: [{ emotions: expect.arrayContaining(expected) }],
+        })
+      })
+    })
+
+    describe('authenticated as author', () => {
+      let headers
+      beforeEach(async () => {
+        headers = await login(authorParams)
+        client = new GraphQLClient(host, { headers })
+      })
+
+      it('adds an emotion to the post', async () => {
+        const expected = {
+          AddPostEmotions: {
+            from: addPostEmotionsVariables.from,
+            to: addPostEmotionsVariables.to,
+            emotion: 'happy',
+          },
+        }
+        await expect(
+          client.request(addPostEmotionsMutation, addPostEmotionsVariables),
+        ).resolves.toEqual(expected)
+      })
+    })
+  })
+
+  describe('RemovePostEmotions', () => {
+    let removePostEmotionsVariables
+    const removePostEmotionsMutation = `
+      mutation($from: _UserInput!, $to: _PostInput!, $data: _EMOTEDInput!) {
+        RemovePostEmotions(from: $from, to: $to, data: $data)
+      }
+    `
+    beforeEach(() => {
+      removePostEmotionsVariables = {
+        from: { id: authorParams.id },
+        to: { id: 'p1376' },
+        data: { emotion: 'cry' },
+      }
+    })
+
+    describe('unauthenticated', () => {
+      it('throws authorization error', async () => {
+        client = new GraphQLClient(host)
+        await expect(
+          client.request(removePostEmotionsMutation, removePostEmotionsVariables),
+        ).rejects.toThrow('Not Authorised')
+      })
+    })
+
+    describe('authenticated', () => {
+      let headers
+      beforeEach(async () => {
+        headers = await login(authorParams)
+        client = new GraphQLClient(host, { headers })
+        await factory.emote({
+          from: authorParams.id,
+          to: 'p1376',
+          data: 'cry',
+        })
+        await factory.emote({
+          from: authorParams.id,
+          to: 'p1376',
+          data: 'happy',
+        })
+      })
+
+      describe('but not the emoter', () => {
```


I love the word `emoter `! 😍

File: webapp/locales/de.json [hunk 2]
```diff
@@ -420,6 +420,14 @@
...
     "categories": {
       "infoSelectedNoOfMaxCategories": "{chosen} von {max} Kategorien ausgewählt"
+    },
+    "emotions-label": {
+      "funny": "Lustig",
+      "happy": "Glücklich",
+      "surprised": "Erstaunt",
+      "cry": "Zum Weinen",
+      "angry": "Verärgert",
+      "emoted": "angegeben"
```
Comment (Tirokk) [inline webapp/locales/de.json:430]: > [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-08-05T14:48:32Z" title="Monday, August 5th 2019, 4:48:32 pm +02:00">Aug 5, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -415,6 +415,14 @@
     "languageSelectLabel": "Sprache",
     "categories": {
       "infoSelectedNoOfMaxCategories": "{chosen} von {max} Kategorien ausgewählt"
+    },
+    "emotions-label": {
+      "funny": "Lustig",
+      "happy": "Glücklich",
+      "surprised": "Erstaunt",
+      "cry": "Zum Weinen",
+      "angry": "Verärgert",
+      "emoted": "emotiert"
```


The word `emotiert` just doesn’t exist in german, as far as I know. And I even don’t know anything similar.
I would set the word to `angegeben`.

File: webapp/pages/post/_id/_slug/index.spec.js [hunk 3]
```diff
@@ -31,7 +31,7 @@ describe('PostSlug', () => {
       $filters: {
         truncate: a => a,
       },
-      // If you mocking router, than don't use VueRouter with lacalVue: https://vue-test-utils.vuejs.org/guides/using-with-vue-router.html
+      // If you are mocking the router, then don't use VueRouter with localVue: https://vue-test-utils.vuejs.org/guides/using-with-vue-router.html
```
Comment (Tirokk) [inline webapp/pages/post/_id/_slug/index.spec.js:34]: > [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-08-05T14:50:18Z" title="Monday, August 5th 2019, 4:50:18 pm +02:00">Aug 5, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -31,7 +31,7 @@ describe('PostSlug', () => {
       $filters: {
         truncate: a => a,
       },
-      // If you mocking router, than don't use VueRouter with lacalVue: https://vue-test-utils.vuejs.org/guides/using-with-vue-router.html
+      // If you are mocking the router, then don't use VueRouter with localVue: https://vue-test-utils.vuejs.org/guides/using-with-vue-router.html
```


That’s an unrelated file !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
😂

File: backend/src/schema/types/type/Post.gql [hunk 4]
```diff
@@ -89,4 +91,11 @@ type Mutation {
     language: String
     categoryIds: [ID]
   ): Post
+  AddPostEmotions(to: _PostInput!, data: _EMOTEDInput!): EMOTED
+  RemovePostEmotions(to: _PostInput!, data: _EMOTEDInput!): EMOTED
+}
+
+type Query {
```
Comment (Tirokk) [inline backend/src/schema/types/type/Post.gql:98]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-06T20:19:48Z" title="Tuesday, August 6th 2019, 10:19:48 pm +02:00">Aug 6, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -89,4 +91,10 @@ type Mutation {
     language: String
     categoryIds: [ID]
   ): Post
+  RemovePostEmotions(from: _UserInput!, to: _PostInput!, data: _EMOTEDInput!): Boolean!
+}
+
+type Query {
+  postsEmotionsCountByEmotion(postId: ID!, data: _EMOTEDInput!): Int!
```


~I don't like STRANGECase, but I'm not going to request changes because of that - but you and everyone else are welcome to change it in the next PR :wink:~
Found improvements and would like to request changes - so you have the option to implement it in this PR :gift: 

Don't use STRANGECase, pleASE

File: backend/src/middleware/permissionsMiddleware.js [hunk 5]
```diff
@@ -157,6 +157,8 @@ const permissions = shield(
       User: or(noEmailFilter, isAdmin),
       isLoggedIn: allow,
       Badge: allow,
+      PostsEmotionsCountByEmotion: allow,
+      PostsEmotionsByCurrentUser: allow,
     },
     Mutation: {
```
Comment (Tirokk) [inline backend/src/middleware/permissionsMiddleware.js:163]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-06T20:23:00Z" title="Tuesday, August 6th 2019, 10:23:00 pm +02:00">Aug 6, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -27,6 +27,12 @@ const onlyYourself = rule({
   return context.user.id === args.id
 })
 
+const isMyEmotion = rule({
+  cache: 'no_cache',
+})(async (parent, args, context, info) => {
+  return context.user.id === args.from.id
```


Eww - implement your own custom resolvers and you don't have to check that anymore.

eww

File: backend/src/schema/resolvers/posts.js [hunk 6]
```diff
@@ -70,5 +71,80 @@ export default {
...
+        return {
+          from: { ...record.get('userFrom').properties },
+          to: { ...record.get('postTo').properties },
+          ...record.get('emotedRelation').properties,
+        }
+      })
+      return emoted
+    },
+    RemovePostEmotions: async (object, params, context, resolveInfo) => {
+      const session = context.driver.session()
```
Comment (Tirokk) [inline backend/src/schema/resolvers/posts.js:95]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-06T20:23:34Z" title="Tuesday, August 6th 2019, 10:23:34 pm +02:00">Aug 6, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -70,5 +71,76 @@ export default {
 
       return post.properties
     },
+    AddPostEmotions: async (object, params, context, resolveInfo) => {
+      const session = context.driver.session()
+      const { from, to, data } = params
+      const transactionRes = await session.run(
+        `MATCH (userFrom:User {id: $from.id}), (postTo:Post {id: $to.id})
+        MERGE (userFrom)-[emotedRelation:EMOTED {emotion: $data.emotion}]->(postTo)
+        RETURN userFrom, postTo, emotedRelation`,
+        { from, to, data },
+      )
+      session.close()
+      const [emoted] = transactionRes.records.map(record => {
+        return {
+          from: { id: record.get('userFrom').properties.id },
+          to: { id: record.get('postTo').properties.id },
+          emotion: record.get('emotedRelation').properties.emotion,
+        }
+      })
+      return emoted
+    },
+    RemovePostEmotions: async (object, params, context, resolveInfo) => {
+      const session = context.driver.session()
+      const { from, to, data } = params
```


```suggestion
      const { to, data } = params
      const { id: from } = context.user
```

File: backend/src/seed/factories/index.js [hunk 7]
```diff
@@ -128,6 +128,22 @@ export default function Factory(options = {}) {
       this.lastResponse = await cleanDatabase({ driver: this.neo4jDriver })
       return this
     },
+    async emote({ to, data }) {
```
Comment (Tirokk) [inline backend/src/seed/factories/index.js:131]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-06T20:26:16Z" title="Tuesday, August 6th 2019, 10:26:16 pm +02:00">Aug 6, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -128,6 +128,23 @@ export default function Factory(options = {}) {
       this.lastResponse = await cleanDatabase({ driver: this.neo4jDriver })
       return this
     },
+    async emote({ from, to, data }) {
```


If you use `neode` you wouldn't have to make a HTTP request over the wire.

File: webapp/pages/post/_id/_slug/index.spec.js (hunk 3 above)
Comment (Tirokk) [reply webapp/pages/post/_id/_slug/index.spec.js:34]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-06T20:31:23Z" title="Tuesday, August 6th 2019, 10:31:23 pm +02:00">Aug 6, 2019</time>_

---


!!!111eleven :angry:

Comment (Tirokk) [reply webapp/pages/post/_id/_slug/index.spec.js:34]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-07T04:55:03Z" title="Wednesday, August 7th 2019, 6:55:03 am +02:00">Aug 7, 2019</time>_

---


what is this supposed to mean? hahaha

File: backend/src/schema/types/type/Post.gql (hunk 4 above)
Comment (Tirokk) [reply backend/src/schema/types/type/Post.gql:98]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-07T05:01:07Z" title="Wednesday, August 7th 2019, 7:01:07 am +02:00">Aug 7, 2019</time>_

---


what's STRANGECase? I found this online https://www.artrix.co.uk/whats-on/all-shows/the-strange-case-of-dr-jekyll-and-mr-hyde

the case used in this example is called camel case and is used quite a lot in javascript, do you prefer pascal case?
https://medium.com/better-programming/string-case-styles-camel-pascal-snake-and-kebab-case-981407998841

File: backend/src/middleware/permissionsMiddleware.js (hunk 5 above)
Comment (Tirokk) [reply backend/src/middleware/permissionsMiddleware.js:163]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-07T05:01:20Z" title="Wednesday, August 7th 2019, 7:01:20 am +02:00">Aug 7, 2019</time>_

---


good catch

File: backend/src/schema/resolvers/posts.js (hunk 6 above)
Comment (Tirokk) [reply backend/src/schema/resolvers/posts.js:95]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-08-07T05:03:24Z" title="Wednesday, August 7th 2019, 7:03:24 am +02:00">Aug 7, 2019</time>_

---


didn't think of that, but like it!!

File: webapp/pages/post/_id/_slug/index.spec.js (hunk 3 above)
Comment (Tirokk) [reply webapp/pages/post/_id/_slug/index.spec.js:34]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-08T21:21:07Z" title="Thursday, August 8th 2019, 11:21:07 pm +02:00">Aug 8, 2019</time>_

---


When people get really angry, they use a lot of exclamation marks and don't manage to keep the shift key pressed!!!11

**Review by Mogge — CHANGES_REQUESTED** [review]
Wow @mattwr18 ,

this is quite a stack of work you did again !!! 💫🚀
Glad to have you around and see again that you did a wonderful job! 😍
Very nice and detailed tests as well …

I found a functioning problem, where if you set emotions on a post:
Bildschirmfoto 2019-08-05 um 16 08 02

… go than to the main page and click there on the post again:
Bildschirmfoto 2019-08-05 um 16 08 20

… the counts of the specific emotions are incorrect displayed.

**Review by Mogge — APPROVED** [review]
Yeah. Now it works, my dear !!!

Great work @mattwr18 , my big emoter !!! Yuhuhu 🚀💫
We getting even closer to the release. 🚀🚀🚀😍

The party comes … 🕺🏼💃🏻❤️

Pipi Langstrumpf and Friends - Kiffing

**Review by Mogge — CHANGES_REQUESTED** [review]
good start

Looks great, I tested it out and it feels good! However, I think you could do some refactoring. Especially the backend should not fall back to `neo4j-graphql-js` anymore. Refactoring the resolvers is hard work and at least we should not add any more legacy resolvers.

**Review by Mogge — APPROVED** [review]
I hope your PR will help Mister Spock along !!! 🖖🏼
At least he can express himself by a machine. 🚀🤔

Thanks for your work again @mattwr18 👍🏼🌟

I am not capable of that emotions

**Review by Mogge — APPROVED** [review]
emotional
